### PR TITLE
[#433] add output mapping

### PR DESCRIPTION
**Pull request by NikkyAI**
this adds support for https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs

specifically adding mapping of a specifc steps output to the output of the job
and accessing the outputs of the job from other jobs (assuming `needs` is also setup)

### CODE REVIEW

File: library/src/test/kotlin/it/krzeminski/githubactions/IntegrationTest.kt [hunk 1]
```diff
@@ -593,6 +593,83 @@ class IntegrationTest : FunSpec({
...
+        ) {
+            val setOutputJob = job(
+                id = "set_output",
+                runsOn = RunnerType.UbuntuLatest,
+            ) {
+                run(
+                    name = "set output",
+                    command = """echo "::set-output name=test::value"""",
+                )
+                    .withOutputMapping("test", "test")
```
Comment (krzema12) [inline library/src/test/kotlin/it/krzeminski/githubactions/IntegrationTest.kt]: Regarding the API, I'm thinking about something safer and in line with what we have for step outputs: . Not properly type-safe yet because of the strings, but at least output names would be defined once:

```kotlin
workflow(
    name = "Test workflow",
    on = listOf(Push()),
    sourceFile = gitRootDir.resolve(".github/workflows/some_workflow.main.kts"),
) {
    // define model for job's output. Use it to set the output in job no. 1 and retrieve it in job no. 2
    data class SetOutputJobOutputs(
        val test: String,
        val pythonVersion: String,
    )

    val setOutputJob = job<SetOutputJobOutputs>(
        id = "set_output",
        runsOn = RunnerType.UbuntuLatest,
    ) {
        val setupPython = uses(SetupPythonV4())
        // Use existing mechainsm to retrieve the step output, and new API to set job output
        // I'm not sure if we'd need to make the SetOutputJobOutputs fields mutable, or maybe use some
        // delegated properties magic, but such syntax would be really cool and convenient.
        // I assume that jobOutput is available in the receiver of this job lambda.
        jobOutputs.pythonVersion = setupPython.outputs.pythonVersion

        run(
            name = "set output",
            command = """echo "::set-output name=test::value"""",
        )
        // TODO: there's no support for step outputs for the custom command step,
        // we may need to add it first
    }

    job(
        id = "use_output",
        runsOn = RunnerType.UbuntuLatest,
        needs = listOf(setOutputJob),
    ) {
        run(
            name = "use output test",
            // Use it similarly like step outputs
            command = """echo ${expr { setOutputJob.outputs.test }}""",
        )
        run(
            name = "use output pythonversion",
            command = """echo ${expr { setOutputJob.outputs.pythonVersion }}""",
        )
    }
}

```

Comment (NikkyAI) [reply library/src/test/kotlin/it/krzeminski/githubactions/IntegrationTest.kt]: tempting to make Job have a type parameter, will probably have to make sure to make it with mutable variables

or something like this maybe ?

```
class SetOutputJobOutputs {
    var pythonVersion by Outputs.register()
}
```

Comment (NikkyAI) [reply library/src/test/kotlin/it/krzeminski/githubactions/IntegrationTest.kt]: i think i can approximate the desired api to some degree

added a `ComandStepWithOutput<T>`  and extension function on `CommandStep`  to convert them (and add the outputs)

generics on `Job` affect a lot of things, i opted for making their default output type Unit  to avoid duplicating more types 
 
and we can make it even closer by adding a overload to` fun expr(block: () -> String)): String`  that accepts non string in the block and converts them to the references (but the typesystem did not like my approach, so i skipped that for now)

Comment (NikkyAI) [reply library/src/test/kotlin/it/krzeminski/githubactions/IntegrationTest.kt]: i managed to get the API very close to your desired form i think

```kotlin
val setOutputJob = job(
    id = "set_output",
    runsOn = RunnerType.UbuntuLatest,
    outputs = object : JobOutputs() {
        var pythonVersion: String by createOutput()
        var bar: String by createOutput()
    }
) {
    val commandStepWithOutput = run(
        name = "set output",
        command = """echo "::set-output name=foo::baz"""",
    ).withOutputs(object : StepOutputs() {
        val foo by property()
    })
    jobOutputs.bar = commandStepWithOutput.outputs.foo

    val setupPython = uses(SetupPythonV4())
    jobOutputs.pythonVersion = setupPython.outputs.pythonVersion
}

job(
    id = "use_output",
    runsOn = RunnerType.UbuntuLatest,
    needs = listOf(setOutputJob),
) {
    run(
        name = "use output test",
        command = """echo ${expr { setOutputJob.jobOutputs.bar }}""",
    )
    run(
        name = "use output pythonversion",
        command = """echo ${expr { setOutputJob.jobOutputs.pythonVersion }}""",
    )
}
```

PS: delegates can be a pain in the ass

File: library/src/test/kotlin/it/krzeminski/githubactions/IntegrationTest.kt [hunk 2]
```diff
@@ -593,6 +595,85 @@ class IntegrationTest : FunSpec({
 
         """.trimIndent()
     }
+
+    test("outputs - access outputs across job") {
```
Comment (krzema12) [inline library/src/test/kotlin/it/krzeminski/githubactions/IntegrationTest.kt:599]: Before we dig into the implementation details, let's first make this test runnable on GitHub. See other tests in this file with `testRanWithGitHub`. It will generate a YAML in .github/workflows which you have to check in, and then it will be ran by GitHub on next push to your branch.

File: library/src/main/kotlin/it/krzeminski/githubactions/dsl/JobBuilder.kt [hunk 3]
```diff
@@ -175,5 +180,24 @@ class JobBuilder(
         return newStep
     }
 
+    fun <T : StepOutputs> CommandStep.withOutputs(
```
Comment (krzema12) [inline library/src/main/kotlin/it/krzeminski/githubactions/dsl/JobBuilder.kt:183]: Adding command step with outputs seems out of scope of adding job outputs. Could you post a separate, smaller PR that adds **only** command step with outputs? We'll discuss its API there.

**Review by krzema12 — CHANGES_REQUESTED** [review]

File: library/src/main/kotlin/it/krzeminski/githubactions/dsl/JobBuilder.kt (hunk 3 above)
Comment (NikkyAI) [reply library/src/main/kotlin/it/krzeminski/githubactions/dsl/JobBuilder.kt:183]: yes, i will remove it from this PR, it is probably not even necessary for my purpose since i can use github-script easier..

File: library/src/main/kotlin/it/krzeminski/githubactions/domain/StepOutputs.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package it.krzeminski.githubactions.domain
+
+import kotlin.properties.ReadOnlyProperty
+
+open class StepOutputs {
```
Comment (krzema12) [inline library/src/main/kotlin/it/krzeminski/githubactions/domain/StepOutputs.kt:5]: Looks like leftover after the previous extended PR, can you remove this file?

File: library/src/main/kotlin/it/krzeminski/githubactions/domain/JobOutputs.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
+package it.krzeminski.githubactions.domain
+
+import it.krzeminski.githubactions.dsl.expressions.expr
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KProperty
+
+open class JobOutputs {
```
Comment (krzema12) [inline library/src/main/kotlin/it/krzeminski/githubactions/domain/JobOutputs.kt:7]: This class has mixed responsibilities: it serves both as a domain class and contains pieces of GitHub YAML serialization logic. Please split it, or let me know if I should do it.

File: library/src/main/kotlin/it/krzeminski/githubactions/domain/Job.kt [hunk 6]
```diff
@@ -4,19 +4,51 @@ import it.krzeminski.githubactions.dsl.HasCustomArguments
...
     val env: LinkedHashMap<String, String> = linkedMapOf(),
     val condition: String? = null,
     val strategyMatrix: Map<String, List<String>>? = null,
     val timeoutMinutes: Int? = null,
     val concurrency: Concurrency? = null,
     override val _customArguments: Map<String, @Contextual Any> = mapOf(),
 ) : HasCustomArguments {
+    companion object {
+        @Suppress("LongParameterList")
+        operator fun invoke(
```
Comment (krzema12) [inline library/src/main/kotlin/it/krzeminski/githubactions/domain/Job.kt:23]: Why is it needed to implement this operator? It looks like you wanted to have a secondary constructor. What's your intent exactly?

**Review by krzema12 — CHANGES_REQUESTED** [review]

File: library/src/main/kotlin/it/krzeminski/githubactions/domain/Job.kt (hunk 6 above)
Comment (NikkyAI) [reply library/src/main/kotlin/it/krzeminski/githubactions/domain/Job.kt:23]: that was my solution to keep the old calling convention without changing how `Job` is constructed all over the codebase while i was still tweaking it..
we can remove that

File: library/src/main/kotlin/it/krzeminski/githubactions/domain/JobOutputs.kt (hunk 5 above)
Comment (NikkyAI) [reply library/src/main/kotlin/it/krzeminski/githubactions/domain/JobOutputs.kt:7]: feel free to tackle that, not sure how you want it split

File: library/src/test/kotlin/it/krzeminski/githubactions/yaml/JobsToYamlTest.kt [hunk 7]
```diff
@@ -283,6 +292,7 @@ class JobsToYamlTest : DescribeSpec({
                         command = "echo 'test!'",
                     ),
                 ),
+                outputs = JobOutputs.EMPTY,
```
Comment (krzema12) [inline library/src/test/kotlin/it/krzeminski/githubactions/yaml/JobsToYamlTest.kt:295]: TODO for myself: add a test case for non-empty outputs.

File: library/src/main/kotlin/it/krzeminski/githubactions/domain/JobOutputs.kt [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
```
Comment (krzema12) [inline library/src/main/kotlin/it/krzeminski/githubactions/domain/JobOutputs.kt:7]: TODO for myself: try to split it so that `Ref` goes into `dsl` package, and `domain` contains only what's needed.

**Review by krzema12 — APPROVED** [review]

**Comment by krzema12** [comment]
@NikkyAI thanks a lot for this contribution! :guitar: :rocket:

File: library/src/main/kotlin/it/krzeminski/githubactions/domain/JobOutputs.kt (hunk 5 above)
Comment (krzema12) [reply library/src/main/kotlin/it/krzeminski/githubactions/domain/JobOutputs.kt:7]: I gave it a shot, but it's not that easy. I decided to leave it as is and see how it works in practice.
